keps/sig-storage/1845-prioritization-on-volume-capacity/README.md

```diff
@@ -0,0 +1,505 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+With [volume topology-aware
+scheduling](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md),
+the scheduler will filter nodes considering volume topology constraints.
+Unbound PVCs will be bound with the smallest fit PVs in a topology, however, the
+scheduler does not try to prioritize nodes in different topologies based on
+volume capacity. This leads to a waste of resources which we should avoid.
+
+This KEP continues the work proposed [here](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md#priority).
+
+## Motivation
+
+Because we didn't take volume capacity into account when scheduling a pod that
+can run in multiple topologies. A large PV may be used by a PVC with a small
+capacity request, even if there are many suitable small PVs in other
+topologies. PVCs with a large capacity request may not find feasible PVs to use
+if too many large PVs are consumed by PVCs with a small capacity request.
+
+Like CPU, Memory resources, the scheduler should take volume capacity into
+account in scheduling pods to ensure balanced resource usage.
+
+### Goals
+
+- Nodes of topology in which volumes with close capacity are preferred in
+  scheduling
+- Support arbitrary PV topology domain (e.g. hostname, zone, rack)
+
+### Non-Goals
+
+- Prioritizing nodes based on the total available capacity of static provisioned
+  PVs
+- Prioritizing nodes based on the total available number of static provisioned
+  PVs
+- Prioritizing nodes based on the total available capacity for dynamic
+  provisioning PVCs
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I need to prepare some local PVs with different
+storage capacities (e.g. 10Gi, 100Gi, 1Ti) as different workloads may have
+different storage requirements.
+
+A PV with 1Ti capacity on one node has the same chance as a PV on another node
+with 10Gi capacity to be bound with a 10Gi request PVC in scheduling. If nodes
+of topology in which volumes with close capacity are always preferred. We can
+achieve better storage resource usage.
+
+See https://github.com/kubernetes/kubernetes/issues/83323.
+
+## Design Details
+
+We have a scheduler plugin called
+[VolumeBinding](https://github.com/kubernetes/kubernetes/tree/master/pkg/scheduler/framework/plugins/volumebinding)
+to filter the nodes based on volume topology constraints. After the filter
+phase, there is a reduced set of nodes that can fit a pod. We can implement
+`Score` interfaces to score nodes based on assumed PVs on each node.
+
+The score is calculated in two steps:
+
+1) First, we group all PVCs by storage class. For PVCs in each storage class,
```

Comment:
I've replaced `all PVCs` with `pod's PVCs`.

yes, they're fixed, but it's possible for a pod to reference multiple PVCs and each of them may request a different kind of storage. so multiple storage classes may be involved.

I think it's better to calculate for each class, like CPU, memory, then we can give different weights to different classes.

For example, given a cluster which has two kinds of local storage classes: `hdd` and `ssd`. A workload may request two disks, one from `hdd` and one from `ssd`. As the `ssd` storage resource may be relatively scarce, the administrator can configure a large weight for `ssd` and hope it's more balanced.

This applies to any other storage with topology constraints.

Reply:
It's clear to me now. Thanks for the explanation.

Another question: suppose the pod requests for just one pvc, if a node doesn't have static provisioned PV, but it's eligible to provision PV dynamically, should this node be scored **0**, or **MAX**? Both are reasonable, but we have to make a decision -  **0** is b/c we have other available PVs, **MAX** is b/c dynamic provisioning will provision the exact size PV (correct me if I'm wrong) requested by the pod.

Reply:
**0** IMO. No volumes are involved in the scoring.

Reply:
Just a followup on Wei's question: a node will pass the filtering phase because "eligible to provision PV dynamically" only make sense in WaitForFirstConsumer case, right?

Reply:
> Just a followup on Wei's question: a node will pass the filtering phase because "eligible to provision PV dynamically" only make sense in WaitForFirstConsumer case, right?

yes

only in the WaitForFirstConsumer case, the PV controller will wait for the scheduler (VolumeBinding) to 1) pre-bind a PV (static binding) 2) choose a node for the dynamic provisioning.

in Immediate mode, the PV controller will find matching PVs for PVC immediately, if not found, try to provision a PV if the storage class has a dynamic provisioner set (assuming all nodes can access the provisioned PV).

Reply:
> in Immediate mode, the PV controller will find matching PVs for PVC immediately, if not found, try to provision a PV if the storage class has a dynamic provisioner set (assuming all nodes can access the provisioned PV).

This is happening outside the scheduler though, right? If the PV wasn't provisioned by the time the scheduler assessed the pod, then this isn't relevant, correct?